### ISSUE TITLE
Hidden constant variables not taking effect when updated

### ISSUE THREAD

**Issue by lbatalha** [grafana/grafana#20600]
**What happened**:
When hidden constant variables are updated, their new value is not used correctly.
**What you expected to happen**:
Updating a hidden variable manually should apply this new value in places where it is used.

**How to reproduce it (as minimally and precisely as possible)**:
To reproduce:
- Create a new constant variable `var` and make it not be hidden for now, with value X
- Create a new (any) graph and add the variable to the title as `[[var]]`
- Notice that the graph title should now contain value X
- Make `var` hidden (`Hidden: Variable`)
- Notice that variable dropdown has disappeared from dashboard but graph title still contains X
- Change the variable value from X to Y, making sure to click update as usual
- Return to the dashboard to see that the graph title now contains Y
- Save the dashboard (notice that "save current variables is not shown)
- Refresh the page or open the dashboard in a new tab
- Notice that the graph title now still says X
- Notice that in the variable edit screen, the value says Y
- Clicking Update again will make this work until the dashboard is accessed again.

To "Fix":
- Make the variable not be hidden
- Back at the dashboard, click the value dropdown for `var`, notice that the only option is Y, but the current value shown is X
- Select Y from the dropdown
- Save Dashboard (check "save current variables")
- You can now hide the variable and Y is correctly shown everywhere, until you decide to change it to Z

**Anything else we need to know?**:
Have not tested with non-constant variables.
The issue is not exclusive to panel titles, also bugged in the query (tested with prometheus queries)
I think the main issue is the "save current variables" option not showing up when saving the dashboard, when variables are changed from the dashboard settings page.

**Environment**:
- Grafana version: 6.4.3 
- Data source type & version: CentOS 7 - Linux/AMD64 official docker image
- OS Grafana is installed on: N/A
- User OS & Browser: Chrome or Firefox, MacOS Mojave or Debian
- Grafana plugins: Any (base install)
- Others: N/A

**Comment by papagian** [comment]
I have set `X=20` and `Y=30` and if `Save current variables` is disabled the resulted JSON in the request is:
```
{ 
   "templating":{ 
      "list":[ 
         { 
            "current":{ 
               "text":"20",
               "value":"20"
            },
            "hide":2,
            "label":null,
            "name":"var",
            "options":[ 
               { 
                  "selected":true,
                  "text":"30",
                  "value":"30"
               }
            ],
            "query":"30",
            "skipUrlSync":false,
            "type":"constant"
         }
      ]
   }
}```

**Comment by lbatalha** [comment]
Hey,
I am now having trouble replicating this correctly as well (at least the original problem)
But I have found another way to replicate it temporarily.
When a variable is create or deleted, and another variable's value is modified, the "Save variables" toggle is not shown as well.
Here is a video replicating:
https://youtu.be/hGmFgjM9ndU

The steps in plain text to replicate:
- Have variable X set to 1 and dashboard fully saved
- Create any other variable
- Update value of variable X to 10
- Click Save Dashboard, notice the "Save variables" toggle switch is missing
- On the dashboard, notice variable in graphs is set to 10
- Reload page
- Notice variable in graphs is set to 1
- Notice variable is defined as 10 in the variables config page

The same issue also exists if you forget to toggle "save variables" once, and will be broken until you change the value again.

The original reported problem is not the above, however many times the value changed of a hidden variable, the toggle to "save variables" never appeared and the templated value never persisted accross reloads.
I believe this is some sort of edge case related to the behaviour shown above, but I have been unable to get reliable steps to reproduce as of yet

I will keep trying, and will record again if I manage to get the edge case.

**Comment by CaffeinatedAndroid** [comment]
Also experienced this. I have a hidden constant variable with an updated value not being read: https://github.com/ryantxu/ajax-panel/issues/62

A weird workaround, of sorts, to get Grafana to realise the variable had been updated was to go to `Dashboard settings` > `Variables`, click on the offending variable, then click on the Input field with the varaible value inside then click off (remove focus from the field). I then went back to the dashboard (_without saving anything!_) and refreshed. Variable was correctly read. I confirmed this by inspecting the network requests. However, once the browser was refreshed it "forgot" the updated value, despite the updated value appearing in the Variable Edit page.

Sounds stupid, but it worked, repeatedly.

For now I simply save the variable as the dashboard default as a temporary, reliable workaround.

**Comment by chinhodado** [comment]
I am experiencing this issue as well. Constants are extremely unreliable, or just flat out doesn't work. Like @CaffeinatedAndroid described, upon loading of a page (e.g. F5) Grafana seems oblivious to what the value of a constant is, and only if I go to settings and 

 - click on the value field and click out to make it lose focus, or 
 - click on the Update button (without changing anything)

will Grafana realizes the value for the constant properly.

Whether the constant is hidden or not doesn't seem to matter.

![image](https://user-images.githubusercontent.com/1388219/90802271-fdff2880-e2e4-11ea-9fb4-33bf6ed5ecc7.png)

**Comment by MarSik** [comment]
@papagian Hi, this is labeled as feature request, but it seems to me it is a bug.

In my case I saved an existing dashboard under a new name and edited the value of a hidden constant. And it kept reverting ti the original value until I made it visible (to make Save variable values visible) and saved the variable values as default when saving the new dashboard. Exactly as described in previous comments.

I am using: Grafana v7.2.1 (72a6c64532)

**Comment by hugohaggmark** [comment]
Hi everyone :)

This issue could be because of some bad UX from our side, I'm afraid. I can see that there's some upvotes and comments that might be related to the original issue, but they might be new issues too. I'd like to sort this out. 

Let's start with the original issue. I can't repro this issue in Grafana 6.4.3. 

But the [video attached by](https://github.com/grafana/grafana/issues/20600#issuecomment-561597061) @lbatalha shows another issue. 

The issue (confirmed on 6.4.3) is:

- Create a new dashboard
- Create a new constant variable named `var` with value `X`
- Create a new (any) graph and add the variable to the title as ${var}
- Notice that the graph title should now contain value X
- Change the variable value from X to Y, clicking `update` as usual
- Return to the dashboard to see that the graph title now contains Y
- Save the dashboard (notice that "save current variables" is not shown) because this is the first time we save the dashboard **[OK]**
- Now add a new constant variable named `var2` with value `Z`
- Now edit the constant variable named `var` with value `A`, making sure to click update as usual
- Now Save the Dashboard ![image](https://user-images.githubusercontent.com/562238/96592682-f100b480-12e8-11eb-994f-e7c39d793dfa.png)
- The `Save current variable` toggle/checkbox should appear (see the pictures below) but it doesn't **[BAD UX??]**
- If you refresh the browser in version 6.4.3 of Grafana the value in the graph will still show the incorrect value `Y` instead of `A`. **[NOT OK]**
- If you refresh the browser in version v7.4.0-pre (8032b43838) of Grafana the value in the graph will show the correct value of `A`. **[OK]**

My suggestions are:
- Upgrade to the latest version of Grafana and see if the issue is still present. I can't repro it using the provided steps, video and comments.
- Anytime you change the value of a variable and want to use that as the __new default value the next time the dashboard is refreshed__ remember to check/toggle the `Save current variable` checkbox/toggle in the Save Dashboard Modal. Otherwise you'll not have a pleasant experience and it won't behave as you expect it. See the dashed markings in the pictures below. This is what I mean by bad UX, we should prevent you (the user) from doing this wrong.

I hope my long comment clarifies this issue somewhat for anyone reading this.

All the best

**Save Dashboard Modal in different versions**
__6.4.3__
![image](https://user-images.githubusercontent.com/562238/96592870-2a392480-12e9-11eb-8244-32e2236320e8.png)

__Grafana v7.4.0-pre (8032b43838)__
![image](https://user-images.githubusercontent.com/562238/96592829-1a214500-12e9-11eb-87dc-0b1ae9143dc2.png)

**Comment by lbatalha** [comment]
Thanks for disambiguating this issue as it seems some  people might be reporting slightly different issues.
I remember a bit after the time of reporting I was unable to  re-create the issue, and a few months later it happened again.
I have been unable to recreate this since then, especially since I have been using the latest grafana versions.

Testing just now with Grafana  7.1.0, following the steps as before it seems `Save current variable values as dashboard default` does appear as expected.

Given this, I assume this issue can be closed for now?

**Comment by hugohaggmark** [comment]
Thank you for that quick reply @lbatalha but before I close this issue I'd like to wait for some feedback from @CaffeinatedAndroid, @chinhodado and @MarSik. 

Your comments also mention issues with constant variables, I'd love to get some more repro steps from you before we can close this issue.

**Comment by chinhodado** [comment]
I just updated to 7.2.2 (from 7.1.3) and the issue as I described above is still there:

Here is a summary of the setup:

I have 2 variables defined, `test` and `test2`.
 -  `test` is a constant 
 - `test2` is a prometheus query that relies on `test`:
![image](https://user-images.githubusercontent.com/1388219/96759686-63839a00-13a6-11eb-9cec-5c3a2ede8e9d.png)
![image](https://user-images.githubusercontent.com/1388219/96759750-75653d00-13a6-11eb-809b-ac8237fff32a.png)
![image](https://user-images.githubusercontent.com/1388219/96759799-831ac280-13a6-11eb-85d0-8e9ea8f26be3.png)

The dashboard is properly saved. When I refresh the page (F5) and go to edit `test2`, I see that it has no value as in the third screenshot above.

However, if I now go back to Variables page, go to edit `test`, and either 

 - click on the Value field, or
 - click on the Update button

without changing any value in any field, then go back to edit `test2` (all this without refreshing the page), then the values gets populated/refreshed properly:

![image](https://user-images.githubusercontent.com/1388219/96763634-a8f49700-13a7-11eb-9fde-a752fbbfafcb.png)

**Comment by hugohaggmark** [comment]
Thank you for adding some repro steps @chinhodado! I think when you saved your dashboard you forgot to check the __Save current variables__ checkbox/toggle. To prove my assumption is correct, open your JSON Model:
![image](https://user-images.githubusercontent.com/562238/96824784-f1f02e00-142f-11eb-955b-46bf600d2a54.png)
 
Find the JSON that represents your variable `test` it will contain something like the JSON below (I've removed a lot of props):
```json
{
    "current": {
      "text": "This is where you've another value that is not od-jetty",
      "value": "This is where you've another value that is not od-jetty"
    },
    "name": "var",
    "query": "od-jetty",
    "type": "constant"
  }
```

If I'm right then current value `"current": {
      "text": "This is where you've another value that is not od-jetty",
      "value": "This is where you've another value that is not od-jetty"
    },` is something different then `od-jetty`.

If I'm wrong, could you please share your JSON for both variables?

I've created a feature request that will address some of this confusion in #28428 

All the best

**Comment by chinhodado** [comment]
Thanks @hugohaggmark, that indeed seems to be the issue.

Original dashboard:
```
      {
        "current": {
          "selected": false,
          "text": "jetty",
          "value": "jetty"
        },
        "hide": 0,
        "label": "test",
        "name": "test",
        "options": [
          {
            "selected": true,
            "text": "od-jetty",
            "value": "od-jetty"
          }
        ],
        "query": "od-jetty",
        "skipUrlSync": false,
        "type": "constant"
      },
```

After checking the "Save current variable values" checkbox:

```
      {
        "current": {
          "selected": false,
          "text": "od-jetty",
          "value": "od-jetty"
        },
        "hide": 0,
        "label": "test",
        "name": "test",
        "options": [
          {
            "selected": true,
            "text": "od-jetty",
            "value": "od-jetty"
          }
        ],
        "query": "od-jetty",
        "skipUrlSync": false,
        "type": "constant"
      },
```

I'm still not 100% sure at which step earlier I did not check that checkbox, but I think it's very bad UX:

 - The checkbox plus the text is fairly unremarkable, and they sometimes show up, sometimes doesn't, on a save dialog that most people will just click save on muscle memory without even looking at the text. This is incredibly easy to miss.
 - Why isn't that checkbox checked by default? If you go into the setting and change something, then **save**, it's more likely that you want to keep those changes rather than discarding them. Otherwise you would have just discarded those changes and not save in the first place.

**Comment by hugohaggmark** [comment]
> Thanks @hugohaggmark, that indeed seems to be the issue.

_I'm very glad I could help out :)_

> * The checkbox plus the text is fairly unremarkable, and they sometimes show up, sometimes not, on a save dialog that most people will just click save on muscle memory without even looking at the text. This is incredibly easy to miss.

_Yes, it is, and apparently there is research that states that check boxes are not optimal to use in these cases._

> * Why isn't that checkbox checked by default? If you go into the setting and change something, then **save**, it's more likely that you want to keep those changes rather than discarding them. Otherwise you would have just discarded those changes and not save in the first place.

_This is a lot trickier, for Constant variables and all hidden variables this should probably be the default behavior. But for other variable types you don't want that behavior, let's take a Query type, for instance. [You might have many values in the drop down so you want to have one of those values as the dashboard default for all users. But when you're working with it, you'd like to change the value without it becoming the new default.](https://github.com/grafana/grafana/issues/28428#issuecomment-713530891)_

I hope this clarifies your questions @chinhodado, we're looking into this and hopefully with the help from the community we'll get this working better.

**Comment by hugohaggmark** [comment]
I'm going to close this issue because this seems to be due to challenging UX and we've opened another issue #28428 that should solve this. We can always reopen this if there is more proof or repro steps. Thank you all for helping out untangle this issue.

All the best
